### ISSUE TITLE
[feature] Multiple Grids in one view

### ISSUE THREAD

**Issue by jmzagorski** [ghiscoding/aurelia-slickgrid#36]
**I'm submitting a feature request**

* **Framework Version:**
2.0 potentially or 1.x if not breaking

**Please tell us about your environment:**
* **Operating System:**
All

* **Browser:**
all

* **Language:**
TypeScript

**Current behavior:**
No supported since services are singletons and hold onto grid state

* **What is the expected behavior?**
Services can act on multiple grids

* **What is the motivation / use case for changing the behavior?**

Copying over the discussion from https://github.com/ghiscoding/aurelia-slickgrid/pull/33 to keep track of progress on this:

Creating a singleton stateful service has some drawbacks, and we discussed one of those yesterday with multiple grids on the same page. I can think of 3 options if you wanted to change your services, both of which would fix the multiple grid problem because it would not have one state.

1. Inject all grid services as transient. This means every slickgrid custom element will get an new instance of the service and work with just with that grid. The downside of this is that these services cannot be used in view models since a new instance is created every time it is injected. However, it would fix the multiple grids on a page issue

2. Use an object on all services to keep track of which grids are passed into init by either using the object itself or the grid id. This could get complicated because every method call to a service would have to pass some key so the service knows which grid to operaton on. Since you would make to pass in a key anyway as an additional argument for this to work, number 3 would be a better choice.

3. Make a stateless services. By no state I mean they do not have properties for the grid or anything the custom element created at runtime, they just perform a service on the grid, dataview, columns or whatever is passed into their methods. For example this line would pass the grid and row number as parameters. Users can still use your service in their view models because you expose grid objects anyway and if a method required the grid they could pass it. This would allow multiple grids to be rendered on the same page too.

After typing all of that, number 1 would be the least amount of change, but break use of your services outside of the custom element. I would not recommend number 2. Number 3 is another good option because services should be stateless, but would required a rewrite of all service methods requiring a grid object.

number 3 would have to be in version 2 and most likely they get the grid from some event or leave it exposed as a bindable.

I think number 1 is just as good and didn't want to rule it out. It potentially would not add any breaking changes while fixing the multiple grid problem (i think). We would need to think of a way to expose your service methods with 1 or many grids, which is challenging if we inject the services with @transient. Based on the index file that would be quite a few service methods.

**Comment by jmzagorski** [comment]
@ghiscoding i am replying to your comment [here](https://github.com/ghiscoding/aurelia-slickgrid/issues/63) so we can track the conversation here.

If you do not want service singletons then you can add `@transient()` decorator to your service classes or you can register the services in the index file of your plugin with `aurelia.container.registerTransient`

see https://aurelia.io/docs/fundamentals/dependency-injection#explicit-configuration for more info

**Comment by ghiscoding** [comment]
@jmzagorski 
I would rather have your expertise in coding this if possible since you have done some of that in the past. I can learn from your code change. If I understand correctly, the best way would be to use your 2nd option and modify the `index.ts`, I think I would prefer that as it is all in 1 location, if that makes any sense, but again I could be wrong.

**Comment by ghiscoding** [comment]
@jmzagorski 
OMG going fully non-singleton breaks a lot of the functionality that I have put in place. For example the `GridExtraService` which can be used to add an item to the grid is now broken because the Service doesn't have the `grid` and `dataView` references anymore from the outside (my App). 

How would you go around this new issue? Do we need to make it a major version `2.x` and recode how we use these services? I really need some of your advice on this.

**Comment by jmzagorski** [comment]
yup that makes sense. It is a tradeoff really. With `index.ts` you have it all in one place and can tell everything that is transient. You just have to remember to put it in `index.ts`. If you decorate the class it might be easier to remember since it is all in the same file.

As soon as I close my other issues I can start testing that.

**Comment by jmzagorski** [comment]
i'll have to experiment with your services, but yes (as you found out), if you do transient then every time you do `@inject` you get a new instance. If your view models had access to the grids then it would have to run `init` on the service. However, if you do not have access to the grid it would be nearly impossible with transient.

This might have to be a 2.x release. I will do some research on it.

**Comment by ghiscoding** [comment]
Yeah I have a big feeling that this will have to be a Major and will include breaking changes. So I guess that there's no need to postponed a release with what is merged (and a few other commit that I will include in the weekend). I think we need much more time to investigate what is the best approach

**Comment by ghiscoding** [comment]
Just a tought, can we add an Event Aggregator that would a return an object containing all Services reference which we could then call our Services the exact same way as before but instead of coming them from the `import { xService } from 'aurelia-slickgrid;`, it would come instead by the Event Aggregator (it could also be bindable events). This way, the Services usage would be the same, user would just have to change how to import the Services. We won't call them through DI in Constructor though but I don't think that would be a big issue, since all of these Services are imported within the Aurelia-Slickgrid and called via an `init()` function of some kind. Does that make any sense or not?

**Comment by ghiscoding** [comment]
@jmzagorski 
Can you comment on last post when you have a chance, I'd like some thought on how to approach this. I also need multiple grids for work, meaning Angular-Slickgrid, and I think that will have to be addressed quite soon

**Comment by jmzagorski** [comment]
I would have to see an example with that `EventAggregator`. It seems a little hacky, but maybe I don't fully understand. 

How many services do you use outside of the `AureliaSlickgridCustomElement` class?

**Comment by ghiscoding** [comment]
We already do the DI and newing up inside the `Aurelia-Slickgrid` component because we do all the `init()` calls. When creating a grid component, we could use an Event Aggregator (similar to `onGridCreated`) and have it return the grid instance from there, which would contain all services for that grid.

For example, what I saw while changing the Singleton Services to non-Singleton as per the [question](https://stackoverflow.com/questions/50400710/angular-typescript-when-using-multiple-components-1st-component-affect-2nd-c) I asked over on Stack Overflow, is that all Services works great when they are called internally by the Angular-Slickgrid (I say Angular because I was testing that with Angular). A great example is the Grid Menu that has "Export to CSV" and "Remove all Sort", these buttons were working great, but if I try to instead use DI of Export Service and connect it myself to a button that I created in my App, it now breaks because my App doesn't know about this Service anymore (because it's no longer Singleton). If I get that grid instance, I do have access to all it's internal service (non-singleton all become internal services). There's an animated gif that I've put the in the [SO Question](https://stackoverflow.com/questions/50400710/angular-typescript-when-using-multiple-components-1st-component-affect-2nd-c) and that behavior shown in the animation is getting fixed after I change to non-singleton, but I can't call the services outside in the App.

**Comment by ghiscoding** [comment]
I say Event Aggregator because I'm more familiar to them, but yeah `@bindable` is the other option. Event Aggregator as the ability to make it global to the App/Project though, it could be an advantage in some cases.

Example of Event Aggregator used globally by the App.
In our App, I have a grid that if you can click on a row, it will open a sidebar quick view and I need to use the Resizer Service and trigger an auto-resize change, this is better done globally with an Observable and so an Event Aggregator is of better use in that case and a `@bindable` is not sufficient.

**Comment by jmzagorski** [comment]
Okay i better understand now. This would be a good way to decouple your services from the public API. What I mean by that is, consumers of your library are no longer reliant on injecting or creating your actual service class, but instead are dependent on that event payload you emit with the Event Aggregator. You could even go as far as attaching the service class method's to this event payload and not the class itself, which would make it one big payload of method signatures that consumers can call.

**Comment by ghiscoding** [comment]
Whichever grid you decide. I mean if you create 2 Aurelia Grids, you will have 2 instances (1 per grid) and then you decide to which grid you want to add by calling the instance of the grid you want. 

For example, if I create 2 Aurelia Grid in the View 
```html
<aurelia-slickgrid 
  grid-id="grid1" 
  column-definitions.bind="columnDefinitions1" 
  grid-options.bind="gridOptions1" 
  dataset.bind="dataset1"
  on-aurelia-grid-created.bind="aureliaGridReady1($event)"
>
</aurelia-slickgrid>

<aurelia-slickgrid 
  grid-id="grid2" 
  column-definitions.bind="columnDefinitions2" 
  grid-options.bind="gridOptions2" 
  dataset.bind="dataset2"
  on-aurelia-grid-created.bind="aureliaGridReady2($event)"
>
</aurelia-slickgrid>
```

and in your ViewModel 
```javascript
export class GridEditorComponent {
  aureliaGridReady2(aureliaGrid) {
    this.aureliaGrid2 = aureliaGrid;
  }
}

**Comment by ghiscoding** [comment]
Yes it's a possibility, we can expose both (we already do with `onGridCreated` and `ag-on-grid-created` [here](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/aurelia-slickgrid.ts#L155)). I like the Event Aggregator for being available globally, anywhere in my project

**Comment by ghiscoding** [comment]
Ok let's start with option 2. then (`dispatchEvent`) because I would prefer to have some kind of global access to any grids from any components if need be.

For your side note, I'm not sure to understand everything but I would rather not make the Service functions part of the Custom Element, because if I add a public function to the Service, I would also need to add it to the Custom Element which will grow exponentially and might be out of sync if I forget to copy over the function exposed. As for making it look like it's part of SlickGrid, I'm also not sure it's a good idea, I really want to make an obvious distinction between what is from SlickGrid (object or function) and was is from the Custom Element.

**Comment by jmzagorski** [comment]
For my side note: I should have elaborated more.

Once we start supporting multiple grids, there will be little need to injecting these custom services outside of the grid custom element because services will be tied to a grid. So these services really become more of an organizational unit that is part of `AureliaSlickGridCustomElement`. You would not replicate the service methods manually, there are a few alternatives:
      
1. Put the method on the custom element instead of creating the services
2. Use a decorator to iterate over all the methods in all your services and attach them to your custom element at runtime
3. When you emit the event add the methods directly in the event. Calling a method would like ``$event.detail.addItemToDatagrid(newItem);``
4. Or do as you mentioned where it looks like each service is a property on the custom grid? `$event.detail.gridExtraService.addItemToDatagrid(newItem);`

**Comment by ghiscoding** [comment]
@jmzagorski 
I need your help, I refactored the code to have all Services as `transient` and I thought it was working... until I tried creating 2 grids in a same view (Example 10 with row selections). This is working fine in my Angular repo, so I assume that I didn't do the change from Singleton to non-singleton correctly.

There's 2 commits that I made for the `transient` change
- [commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/daf98c16f7e0631c6f67ab81bf801a9bdbe3b3f9) to change `transient`
- [commit](https://github.com/ghiscoding/aurelia-slickgrid/commit/27d1d2b6b8821250b0597079be14b01cc5a14e8d) to add multiple grids in a view

From what is shown below, you can see that the 1st grid is totally not there (it was override by the 2nd grid) and if I sort 2 different columns, if I clear the Sort in 1st grid, it happens in the 2nd grid. So the transient is obviously not working. 

Please help.... the branch is `feature/new-major-2-release`

![2018-05-26_18-11-29](https://user-images.githubusercontent.com/643976/40580642-9fa4469c-6110-11e8-8710-4e8bcbd0f198.gif)

**Comment by jmzagorski** [comment]
@ghiscoding 
I might not be able to look at this for a few days. I was able to render two grids on top of each other, but I had to explicitly set the height in the `gridOptions`.

Ill take a look at the commits as soon as I can

**Comment by ghiscoding** [comment]
@jmzagorski 
you were right about the fixed dimensions, that fixed the 1st grid not showing up. I made a commit that has both grids showing correctly.

However, I still have problems with the Services, they are still Singleton. I think it's because of the Container, which I am not familiar with. I also tried to instead use the `@transient()` inside the class (for example the ExportService), however when I do that it says that the `this._grid` is unknown inside that ExportService. I might have sync issues, the "Export to CSV" (from the hamburger menu) is calling it inside the ControlAndPluginService, which then calls the ExportService, but if you look at Aurelia-Slickgrid.ts, you can see that ControlAndPluginService is created before ExportService, does that matter?

Anyhow, you can pull the branch and you should be able to test that out

**Comment by ghiscoding** [comment]
I think it would probably start working if I create new Service instances inside the Aurelia-Slickgrid custom element but that would not take advantage of DI and I didn't have to do anything special in my Angular repo

**Comment by ghiscoding** [comment]
@jmzagorski 
This issue is the show stopper to resolve before being able to use multiple grids, I kinda need your expertise on this. I don't know what else to try, I suggest you to checkout the branch and try it out, whenever possible for you.

**Comment by jmzagorski** [comment]
I will have time to look it over tomorrow

**Comment by ghiscoding** [comment]
@jmzagorski 
For reference, I just released Angular-Slickgrid with all these changes on a Major version release. You can see the [GitHub demo](https://ghiscoding.github.io/Angular-Slickgrid). There's 2 Examples with multiple grids. :)

**Comment by jmzagorski** [comment]
@ghiscoding 
You were close with the `registerTransient`, but you need to simplify:

```javascript
  aurelia.container.registerTransient(ControlAndPluginService);
  aurelia.container.registerTransient(ExportService);
  aurelia.container.registerTransient(FilterService);
  aurelia.container.registerTransient(GraphqlService);
  aurelia.container.registerTransient(GridEventService);
  aurelia.container.registerTransient(GridService);
  aurelia.container.registerTransient(GridStateService);
  aurelia.container.registerTransient(GridOdataService);
  aurelia.container.registerTransient(GroupingAndColspanService);
  aurelia.container.registerTransient(OdataService);
  aurelia.container.registerTransient(ResizerService);
  aurelia.container.registerTransient(SortService);
```

Also, the services injected into `ControlAndPluginService` will not work because none of them call `init` since `init` is called in `AureliaSlickgridCustomElement`. All of these services will rely on a new fresh grid for this multiple grids feature. We can fallback to newing these services up ourselves in `AureliaSlickGridCustomElement` and inject the grid into the ctor since you are emiting an event with all of the services now anyway. That goes around dependency injection, but you can't do dependency injection with multiple grids when they rely on a custom element instance. I am still looking for alternative solutions

How do your Angular repo know which grid to operate on with the `providers` from your SO question?

**Comment by ghiscoding** [comment]
@jmzagorski 
hmm I also tried without providing a name to the `registerTransient` but that didn't work. 

I was thinking about `new`ing ourselves but I was hoping to not have to do this ( I really don't need to do this in Angular-Slickgrid). In my Angular repo, the only thing I have to do to make them non-singleton is to use `providers: [service1, service2, ...]`, I really don't like the non-expressive Angular words but adding the `providers` basically turn the Service Singleton into a non-singleton for wherever this Service is used (in my case the Angular-Slickgrid component). In other words, it creates 1 instance for that component and I can still use DI in all the Services because they are used within the Angular-Slickgrid component and I also removed the `export` of the Services so that no one can import them anymore in their App, they have to pass by the new instance that I created.

Note that Angular Component is the same as Custom Element in the Aurelia world.

I don't know how to code it, which is why I asked you, but aren't we able to say that the `registerTransient` Container is the Aurelia-Slickgrid Custom Element and anything within that Container (basically all other Services) will automatically have the same instance using DI. Meaning Singleton within the Custom Element but non-singleton outside of the Custom Element. Unless I misunderstood the meaning of DI Container

**Comment by jmzagorski** [comment]
> hmm I also tried without providing a name to the registerTransient but that didn't work.

Providing the same string name as the key to the container, such as `aurelia.container.registerTransient(SERVICE_NAME, ControlAndPluginService);` registers all services together so you can use the `All` resolver to get an array of services, and if you do not use `All`, Aurelia probably falls back to using singletons.

I think I am onto something.  In my view-models I use the decorator `@singleton(true)` so the ctors are not called every time the user navigates back and forth from the view model. After looking at Angular, and from what you said this might be the key:

> Normally, types are auto-registered as singletons in the root container. So, why do we provide this decorator? This decorator allows you to specify true as an argument to indicate that the singleton should be registered not in the root container, but in the immediate container to which the initial request was issued.

I pushed new code to your branch to test

**Comment by ghiscoding** [comment]
oh nice I'll try it out in an hour or so, that look promising :)

**Comment by ghiscoding** [comment]
@jmzagorski 
Thanks, that does seem to work as it should. From Example 10, I tried Sort in both Grids, then tried Clear Sort (from Grid Menu) and also Export to CSV. It all works... sweet 🎉 

I can continue the work, thanks a ton :)
After all, you're probably the biggest user of Aurelia-Slickgrid (I didn't even start my own project with it yet lol) 🥇

**Comment by ghiscoding** [comment]
closed by PR #75
